File: src/ragrep/ingest/extract.py
"""Extract text content from downloaded Slack files (images, PDFs, Office docs)."""

import asyncio
import json
import logging
import time
from pathlib import Path

log = logging.getLogger(__name__)

IMAGE_TYPES = {"png", "jpg", "jpeg", "gif", "webp", "heic", "bmp", "tiff", "svg"}
MIME_MAP = {
    "png": "image/png", "jpg": "image/jpeg", "jpeg": "image/jpeg",
    "gif": "image/gif", "webp": "image/webp", "heic": "image/heic",
    "bmp": "image/bmp", "tiff": "image/tiff", "svg": "image/svg+xml",
}
TEXT_TYPES = {
    "text", "markdown", "python", "javascript", "typescript", "sql",
    "json", "yaml", "xml", "csv", "shell", "go", "rust", "java",
    "c", "cpp", "html", "css", "plain", "email",
}

VISION_PROMPT = (
    "Describe this image in detail. If it contains text (screenshots, documents, "
    "diagrams, error messages, code), transcribe all visible text. If it's a chart "
    "or graph, describe the data and trends. Focus on information content."
)


def extract_pdf(file_path: Path) -> str | None:
    """Extract text from a PDF file."""
    from pypdf import PdfReader

    try:
        reader = PdfReader(file_path)
        pages = []
        for page in reader.pages:
            text = page.extract_text()
            if text:
                pages.append(text.strip())
        content = "\n\n".join(pages)
        return content if len(content) >= 50 else None
    except Exception as e:
        log.debug("PDF extraction failed for %s: %s", file_path.name, e)
        return None


def extract_docx(file_path: Path) -> str | None:
    """Extract text from a DOCX file."""
    from docx import Document

    try:
        doc = Document(file_path)
        paragraphs = [p.text for p in doc.paragraphs if p.text.strip()]
        content = "\n".join(paragraphs)
        return content if len(content) >= 50 else None
    except Exception as e:
        log.debug("DOCX extraction failed for %s: %s", file_path.name, e)
        return None


def extract_xlsx(file_path: Path) -> str | None:
    """Extract text from an XLSX file."""
    from openpyxl import load_workbook

    try:
        wb = load_workbook(file_path, read_only=True, data_only=True)
        lines = []
        for sheet in wb.worksheets:
            lines.append(f"## Sheet: {sheet.title}")
            for row in sheet.iter_rows(values_only=True):
                vals = [str(v) if v is not None else "" for v in row]
                if any(vals):
                    lines.append("\t".join(vals))
        wb.close()
        content = "\n".join(lines)
        return content if len(content) >= 50 else None
    except Exception as e:
        log.debug("XLSX extraction failed for %s: %s", file_path.name, e)
        return None


def extract_pptx(file_path: Path) -> str | None:
    """Extract text from a PPTX file."""
    from pptx import Presentation

    try:
        prs = Presentation(file_path)
        slides = []
        for i, slide in enumerate(prs.slides, 1):
            texts = []
            for shape in slide.shapes:
                if shape.has_text_frame:
                    for para in shape.text_frame.paragraphs:
                        text = para.text.strip()
                        if text:
                            texts.append(text)
            if texts:
                slides.append(f"## Slide {i}\n" + "\n".join(texts))
        content = "\n\n".join(slides)
        return content if len(content) >= 50 else None
    except Exception as e:
        log.debug("PPTX extraction failed for %s: %s", file_path.name, e)
        return None


def extract_text(file_path: Path) -> str | None:
    """Read a text file."""
    try:
        content = file_path.read_text(errors="replace").strip()
        return content if len(content) >= 50 else None
    except Exception as e:
        log.debug("Text read failed for %s: %s", file_path.name, e)
        return None


async def describe_image(
    file_path: Path,
    file_type: str,
    client: "genai.Client",
    model: str,
    sem: asyncio.Semaphore,
) -> str | None:
    """Describe an image using Gemini Vision."""
    from google.genai import types

    mime = MIME_MAP.get(file_type, f"image/{file_type}")
    try:
        data = file_path.read_bytes()
    except Exception as e:
        log.debug("Failed to read image %s: %s", file_path.name, e)
        return None

    try:
        async with sem:
            response = await client.aio.models.generate_content(
                model=model,
                contents=[
                    VISION_PROMPT,
                    types.Part.from_bytes(data=data, mime_type=mime),
                ],
            )
        return response.text if response.text else None
    except Exception as e:
        log.warning("Gemini Vision failed for %s: %s", file_path.name, e)
        return None


async def extract_all(
    files_jsonl: Path,
    raw_dir: Path,
    output_path: Path,
    gemini_api_key: str,
    vision_model: str = "gemini-2.0-flash",
    vision_concurrency: int = 5,
) -> None:
    """Extract content from all downloaded files."""
    from google import genai

    records = []
    with open(files_jsonl) as f:
        for line in f:
            if line.strip():
                records.append(json.loads(line))

    log.info("Loaded %d file records", len(records))

    client = genai.Client(api_key=gemini_api_key)
    sem = asyncio.Semaphore(vision_concurrency)
    start = time.monotonic()

    stats = {"images": 0, "pdfs": 0, "docx": 0, "xlsx": 0, "pptx": 0, "text": 0, "skipped": 0, "failed": 0}
    results: list[dict] = []

    image_records = []
    for rec in records:
        ft = rec.get("file_type", "")
        local_path = raw_dir / rec.get("local_path", "")

        if not local_path.exists():
            stats["skipped"] += 1
            continue

        if ft in IMAGE_TYPES:
            image_records.append(rec)
            continue

        content = None
        method = "text"
        if ft == "pdf":
            content = extract_pdf(local_path)
            method = "pdf"
            stats["pdfs"] += 1
        elif ft == "docx":
            content = extract_docx(local_path)
            method = "docx"
            stats["docx"] += 1
        elif ft in ("xlsx", "xls"):
            content = extract_xlsx(local_path)
            method = "xlsx"
            stats["xlsx"] += 1
        elif ft == "pptx":
            content = extract_pptx(local_path)
            method = "pptx"
            stats["pptx"] += 1
        elif ft in TEXT_TYPES:
            content = extract_text(local_path)
            method = "text"
            stats["text"] += 1
        else:
            stats["skipped"] += 1
            continue

        if content:
            results.append({**rec, "content": content, "extraction_method": method})
        else:
            stats["failed"] += 1

    log.info(
        "Local extraction done: %d PDFs, %d DOCX, %d XLSX, %d PPTX, %d text. %d images pending.",
        stats["pdfs"], stats["docx"], stats["xlsx"], stats["pptx"], stats["text"], len(image_records),
    )

    async def _process_image(rec: dict) -> dict | None:
        ft = rec.get("file_type", "")
        local_path = raw_dir / rec.get("local_path", "")
        content = await describe_image(local_path, ft, client, vision_model, sem)
        if content:
            return {**rec, "content": content, "extraction_method": "vision"}
        return None

    if image_records:
        coros = [_process_image(rec) for rec in image_records]
        done = 0
        for coro in asyncio.as_completed(coros):
            result = await coro
            if result:
                results.append(result)
                stats["images"] += 1
            else:
                stats["failed"] += 1
            done += 1
            if done % 50 == 0:
                elapsed = time.monotonic() - start
                log.info("[%d/%d images] extracted=%d (%.0fs)", done, len(image_records), stats["images"], elapsed)

    with open(output_path, "w") as f:
        for rec in results:
            f.write(json.dumps(rec) + "\n")

    elapsed = time.monotonic() - start
    log.info(
        "Extraction complete: %d records written (%.0fs). "
        "images=%d, pdfs=%d, docx=%d, xlsx=%d, pptx=%d, text=%d, skipped=%d, failed=%d",
        len(results), elapsed,
        stats["images"], stats["pdfs"], stats["docx"], stats["xlsx"],
        stats["pptx"], stats["text"], stats["skipped"], stats["failed"],
    )


def extract(
    files_jsonl: Path,
    raw_dir: Path,
    output_path: Path,
    gemini_api_key: str,
    vision_model: str = "gemini-2.0-flash",
    vision_concurrency: int = 5,
) -> None:
    """Sync entry point."""
    asyncio.run(extract_all(files_jsonl, raw_dir, output_path, gemini_api_key, vision_model, vision_concurrency))

File: rust/src/ingest/extract.rs
//! Extract text content from downloaded files.
//!
//! Rust v1 keeps local reads for text-like files and uses Gemini multimodal
//! HTTP for PDFs, Office documents, and images.

use std::path::{Path, PathBuf};
use std::time::Duration;

use anyhow::{anyhow, bail, Context, Result};
use base64::Engine;
use serde_json::{json, Value};

use crate::ingest::normalize::read_jsonl;
use crate::ingest::scrape::write_jsonl;

const DEFAULT_MODEL: &str = "gemini-2.0-flash";
const REQUEST_TIMEOUT: Duration = Duration::from_secs(120);
const PROMPT: &str = "Extract the useful information from this file. If it contains text, transcribe all meaningful text. If it is an image, describe it in detail and transcribe visible text. If it is a spreadsheet, preserve rows in a readable tab-delimited form. Focus on information content.";

const IMAGE_TYPES: &[&str] = &[
    "png", "jpg", "jpeg", "gif", "webp", "heic", "bmp", "tiff", "svg",
];
const DOC_TYPES: &[&str] = &["pdf", "docx", "doc", "xlsx", "xls", "pptx", "ppt"];
const TEXT_TYPES: &[&str] = &[
    "text",
    "markdown",
    "python",
    "javascript",
    "typescript",
    "sql",
    "json",
    "yaml",
    "xml",
    "csv",
    "shell",
    "go",
    "rust",
    "java",
    "c",
    "cpp",
    "html",
    "css",
    "plain",
    "email",
];

#[derive(Debug, Default)]
struct Stats {
    vision: usize,
    text: usize,
    skipped: usize,
    failed: usize,
}

pub fn extract_all(raw_dir: &Path) -> Result<usize> {
    let files_jsonl = raw_dir.join("files.jsonl");
    let output_path = raw_dir.join("files_extracted.jsonl");
    let records = read_jsonl(&files_jsonl)?;
    if records.is_empty() {
        eprintln!("No file records found at {}", files_jsonl.display());
        write_jsonl::<Value>(&output_path, &[])?;
        return Ok(0);
    }

    let model = std::env::var("GEMINI_VISION_MODEL").unwrap_or_else(|_| DEFAULT_MODEL.to_string());
    let api_key = std::env::var("GEMINI_API_KEY").ok();
    let client = reqwest::blocking::Client::builder()
        .timeout(REQUEST_TIMEOUT)
        .build()
        .context("building Gemini HTTP client")?;

    let mut stats = Stats::default();
    let mut out = Vec::new();

    for mut record in records {
        let file_type = value_str(&record, "file_type").to_lowercase();
        let local_path = raw_dir.join(value_str(&record, "local_path"));
        if !local_path.exists() {
            stats.skipped += 1;
            continue;
        }

        let extracted = if TEXT_TYPES.contains(&file_type.as_str()) {
            stats.text += 1;
            extract_text(&local_path).map(|content| (content, "text".to_string()))
        } else if IMAGE_TYPES.contains(&file_type.as_str())
            || DOC_TYPES.contains(&file_type.as_str())
        {
            let Some(key) = api_key.as_deref() else {
                stats.failed += 1;
                eprintln!(
                    "warning: GEMINI_API_KEY is not set; cannot extract {}",
                    local_path.display()
                );
                continue;
            };
            stats.vision += 1;
            describe_file(&client, key, &model, &local_path, &file_type)
                .map(|content| (content, "vision".to_string()))
        } else {
            stats.skipped += 1;
            continue;
        };

        match extracted {
            Ok((Some(content), method)) => {
                if let Some(obj) = record.as_object_mut() {
                    obj.insert("content".to_string(), Value::String(content));
                    obj.insert("extraction_method".to_string(), Value::String(method));
                    out.push(record);
                }
            }
            Ok((None, _)) => stats.failed += 1,
            Err(err) => {
                stats.failed += 1;
                eprintln!(
                    "warning: extraction failed for {}: {err:#}",
                    local_path.display()
                );
            }
        }
    }

    write_jsonl(&output_path, &out)?;
    eprintln!(
        "Extraction complete: {} records written. vision={}, text={}, skipped={}, failed={}",
        out.len(),
        stats.vision,
        stats.text,
        stats.skipped,
        stats.failed
    );
    Ok(out.len())
}

fn extract_text(path: &Path) -> Result<Option<String>> {
    let content = std::fs::read_to_string(path)
        .with_context(|| format!("reading text file {}", path.display()))?
        .trim()
        .to_string();
    Ok((content.len() >= 50).then_some(content))
}

fn describe_file(
    client: &reqwest::blocking::Client,
    api_key: &str,
    model: &str,
    path: &Path,
    file_type: &str,
) -> Result<Option<String>> {
    let data = std::fs::read(path).with_context(|| format!("reading {}", path.display()))?;
    if data.is_empty() {
        return Ok(None);
    }
    let encoded = base64::engine::general_purpose::STANDARD.encode(data);
    let mime = mime_for(file_type);
    let url = format!(
        "https://generativelanguage.googleapis.com/v1beta/models/{model}:generateContent?key={api_key}"
    );
    let resp = client
        .post(url)
        .json(&json!({
            "contents": [{
                "parts": [
                    { "text": PROMPT },
                    { "inline_data": { "mime_type": mime, "data": encoded } }
                ]
            }]
        }))
        .send()
        .context("sending Gemini extraction request")?;

    if !resp.status().is_success() {
        let status = resp.status();
        let body = resp.text().unwrap_or_default();
        bail!("Gemini extraction failed: HTTP {status} — {body}");
    }

    let payload: Value = resp.json().context("parsing Gemini extraction response")?;
    let text = payload
        .get("candidates")
        .and_then(Value::as_array)
        .and_then(|candidates| candidates.first())
        .and_then(|candidate| candidate.get("content"))
        .and_then(|content| content.get("parts"))
        .and_then(Value::as_array)
        .into_iter()
        .flatten()
        .filter_map(|part| part.get("text").and_then(Value::as_str))
        .collect::<Vec<_>>()
        .join("\n")
        .trim()
        .to_string();

    Ok((text.len() >= 20).then_some(text))
}

fn mime_for(file_type: &str) -> &'static str {
    match file_type {
        "png" => "image/png",
        "jpg" | "jpeg" => "image/jpeg",
        "gif" => "image/gif",
        "webp" => "image/webp",
        "heic" => "image/heic",
        "bmp" => "image/bmp",
        "tiff" => "image/tiff",
        "svg" => "image/svg+xml",
        "pdf" => "application/pdf",
        "docx" => "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
        "doc" => "application/msword",
        "xlsx" => "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        "xls" => "application/vnd.ms-excel",
        "pptx" => "application/vnd.openxmlformats-officedocument.presentationml.presentation",
        "ppt" => "application/vnd.ms-powerpoint",
        _ => "application/octet-stream",
    }
}

fn value_str(record: &Value, key: &str) -> String {
    record
        .get(key)
        .and_then(Value::as_str)
        .unwrap_or("")
        .to_string()
}

#[allow(dead_code)]
fn _path_for(raw_dir: &Path, relative: &str) -> Result<PathBuf> {
    if relative.is_empty() {
        return Err(anyhow!("missing local_path"));
    }
    Ok(raw_dir.join(relative))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn maps_mime_types() {
        assert_eq!(mime_for("pdf"), "application/pdf");
        assert_eq!(mime_for("png"), "image/png");
    }
}
